__main__.py

```diff
@@ -3,81 +3,149 @@
+            data = cast("dict[str, str]", old.copy())
+            out["data"] = data
+        else:
+            out = cast(StateDict, old)
+        return out
+
+    @classmethod
+    def from_dict(cls, state: StateDict | dict[str, Any], /) -> Self:
+        state_dict = cls._migrate(state)
+        data_raw = state_dict["data"]
+        data = {PEP(k): Status(v) for k, v in data_raw.items()}
+        time = dt.datetime.fromisoformat(state_dict["time"])
+        return cls(data=data, time=time)
+
+    def to_dict(self) -> StateDict:
+        return StateDict(
+            version=self.version,
+            data={str(k): v.value for k, v in self._data.items()},
+            time=self.time.isoformat(),
+        )
+
 
-    def dump(self) -> str:
-        return json.dumps(self._data, indent=2)
+def write_latest(state: State) -> None:
+    LATEST.write_text(json.dumps(state.to_dict(), indent=2))
 
 
 def get_last_state() -> State:
     if not LATEST.exists():
-        dprint("&4latest.json not found.", file=sys.stderr)
-        LATEST.write_text(get_current_state().dump())
-        dexit("&2latest.json created. Run the program again.")
+        dprint(f"&4{LATEST.name} not found.", file=sys.stderr)
+        write_latest(get_current_state())
+        dexit(f"&2{LATEST.name} created. Run the program again.")
     with LATEST.open() as f:
-        return State(
-            json.load(f),
-            dt.datetime.fromtimestamp(LATEST.stat().st_mtime, tz=dt.timezone.utc),
-        )
+        return State.from_dict(json.loads(f.read()))
 
 
 def get_current_state() -> State:
     try:
         req = requests.get(URL, timeout=15)
-    except (requests.ReadTimeout):
+        req.raise_for_status()
+    except requests.ReadTimeout:
         dexit("&4Connection timed out.")
-    if req.status_code != 200:
+    except HTTPError:
         dexit(f"API response code: {req.status_code}")
-    return State({k: v["status"] for k, v in json.loads(req.text).items()})
+    return State.from_api(json.loads(req.text))
 
 
-def pretty_delta(a: dt.datetime, b: dt.datetime) -> str:
-    days, seconds = divmod(round((b - a).total_seconds()), 86400)
+def pretty_delta(delta: dt.timedelta) -> str:
+    seconds = round(delta.total_seconds())
+    days, seconds = divmod(seconds, 86400)
     hours, seconds = divmod(seconds, 3600)
     minutes, seconds = divmod(seconds, 60)
 
-    if days > 0:
-        return f"{days} days" if days != 1 else "day"
-    if hours > 0:
-        return f"{hours} hours" if hours != 1 else "hour"
-    if minutes > 0:
-        return f"{minutes} minutes" if minutes != 1 else "minute"
-    return f"{seconds} seconds" if seconds != 1 else "second"
+    if days:
+        return f"{days} days" if days > 1 else "day"
+    if hours:
+        return f"{hours} hours" if hours > 1 else "hour"
+    if minutes:
+        return f"{minutes} minutes" if minutes > 1 else "minute"
+    return f"{seconds} seconds" if seconds > 1 else "second"
```

Comment:
This function now receives a `timedelta`.

I've broken down the first line into two, which makes it slightly easier to read, and also coincidentally makes it so that this entire first block has the same width, which is visually pleasing.

I've tweaked the logic a bit, but it might not work correctly with negative deltas.

There was a case where `seconds` was between 0 and 1, that it'd print "last 0 seconds", and now it prints "last second".

Reply:
"last 0 seconds" was correct but I like the change 👍